### PR TITLE
ui/tracez: display aggregated data on children spans

### PR DESCRIPTION
**Pull request by benbardin**
Return and display spans' children metadata on /debug/tracez page.

Release note: None
Informs: None

New children metadata table rendered:
Screenshot 2023-03-06 at 4 40 20 PM
Screenshot 2023-03-06 at 4 40 14 PM

No children metadata on this span, new section correctly omitted:
Screenshot 2023-02-28 at 2 34 50 PM

### CODE REVIEW

**Review by andreimatei — COMMENTED** [review]
Awesome!
Can we also add the span's logs with this occasion? I'd take them dumped on the page in any way, no polish needed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-:-NPYuJKOC0qA1So9xu__:bmp29gv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @benbardin, and @dhartunian)

___
*[`pkg/server/serverpb/admin.proto` line 1416 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYmtr66H3r_zgn43uI:-NPYmtr66H3r_zgn43uJ:bvofgyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/server/serverpb/admin.proto#L1416)):*
> ```protobuf
>   // not set if current == false.
>   util.tracing.tracingpb.RecordingMode current_recording_mode = 9;
>   map<string, util.tracing.tracingpb.OperationMetadata> children_metadata = 10 [(gogoproto.nullable) = false];
> ```

Please add a comment explaining that the key in the map is an operation name.

Or consider making this an array, and moving `SpanMetadataRow` to this proto (under a different name). The UI seems inconvenienced by this map form, and the whole point of this proto seems to be to make things easy for the UI.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYp4dREc5R6lUIwd3K:-NPYp4dREc5R6lUIwd3L:bnpzn4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L171)):*
> ```XML
>     useState<SortSetting>();
>   const childrenMetadata = span?.children_metadata;
>   const childrenMetadataAsJson = JSON.stringify(childrenMetadata);
> ```

We're creating the json only so that we can use it as a dependency below? Can we not use `childrentMetadata` as a dep directly for some reason? Please add a comment - unless there's some react/typescript usual pattern happening here.

Is the whole memo business worth it? What happens without it? Are renders of this component frequent?
I don't know what I'm talking about, but I suspect that that jsonifying is more expensive than the computation you're memoizing.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYvP61Av8GDUlK7lCN:-NPYvP61Av8GDUlK7lCO:b-qlqts8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L174)):*
> ```XML
>   const childMetadataRows = useMemo(
>     () =>
>       childrenMetadata
> ```

Do we really need the nil/empty check on `childrenMetadata` or would the code get an empty array without it? Simplifying this expression if possible would certainly be good.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYtuId4U5seK1Il8jx:-NPYtuId4U5seK1Il8jy:b1qnir3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L175)):*
> ```XML
>     () =>
>       childrenMetadata
>         ? Object.keys(childrenMetadata).map((name): SpanMetadataRow => {
> ```

I don't know what I'm talking about, but is `Object.keys` really how you're supposed to iterate over a map?
Is there `childrenMetadata.forEach((name, meta) => ...)`, and then you don't need a lookup inside the closure?
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYox1N8hsZBbvkqPK6:-NPYox1O6VzvwoiZX-5K:b-uzjjm3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L185)):*
> ```XML
>           })
>         : [],
>     // eslint-disable-next-line react-hooks/exhaustive-deps
> ```

do such pragmas deserve an explanation?
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYo5FV7iJ_voizVR1Z:-NPYo5FV7iJ_voizVR1_:bi14y94) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L256)):*
> ```XML
>             Aggregated Child Span Metadata
>           </h3>
>           <Loading
> ```

what is this `Loading` business? We looked at `childMetadataRows` above; doesn't that imply everything is already loaded?
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanMetadataTable.tsx` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYt0Ad8XwrVO-lG2oI:-NPYt0Ad8XwrVO-lG2oJ:bqdpmfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanMetadataTable.tsx#L19)):*
> ```XML
> import styles from "../snapshot.module.scss";
> import classNames from "classnames/bind";
> import { TimestampToMoment } from "src/util";
> ```

are `TimestampToMoment`, `ISpanTag` and `RecordingMode` used in this file (I only checked these 3 cause the list seems long to me)? Perhaps I don't know what imports do.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanMetadataTable.tsx` line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYnKro6npDwQaJ2N48:-NPYnKro6npDwQaJ2N49:bkbeii2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanMetadataTable.tsx#L45)):*
> ```XML
>     title: "",
>     cell: row => {
>       return row.containsUnfinished ? (
> ```

When I saw the screenshot, I wondered what the circle means in the context of a summary. Please put a tooltip explaining that *some* of the spans are still alive.

**Review by benbardin — COMMENTED** [review]
We're already printing the raw trace! That's in prod now (though the logs are only available for recorded spans)




> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-:-NPruU8o3822sAGbHWol:b-85ke66)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @andreimatei, and @dhartunian)

___
*[`pkg/server/serverpb/admin.proto` line 1416 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYmtr66H3r_zgn43uI:-NPrucGy8emh-hUneiH8:b-pz10zm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/server/serverpb/admin.proto#L1416)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Please add a comment explaining that the key in the map is an operation name.

Or consider making this an array, and moving `SpanMetadataRow` to this proto (under a different name). The UI seems inconvenienced by this map form, and the whole point of this proto seems to be to make things easy for the UI.
</blockquote></details>

Yes! Done. Now it's an array.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYp4dREc5R6lUIwd3K:-NPZ-Oq76rNmGLZLzCWH:b66gotp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L171)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

We're creating the json only so that we can use it as a dependency below? Can we not use `childrentMetadata` as a dep directly for some reason? Please add a comment - unless there's some react/typescript usual pattern happening here.

Is the whole memo business worth it? What happens without it? Are renders of this component frequent?
I don't know what I'm talking about, but I suspect that that jsonifying is more expensive than the computation you're memoizing.
</blockquote></details>

Yeah, this is "standard react." Basically, the idea is that for some large N, serializing to Json is cheaper than whatever object-level stuff you're doing in the loop _and then rerendering whatever depends on it_. (JSON serialization is hard-coded in the browser, written in C, dumb as heck and super-fast.) In this case I'm instantiating a bunch of objects, which takes some time, but then also rendering them, which takes way more. For just one or two children it wouldn't be worth it, but since we could conceivably have an unlimited number adding this is typical (I think - maybe David could speak more to that).

That said, I like your idea about pushing this to the server, so it will go away.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYvP61Av8GDUlK7lCN:-NPrtezJ2WK1ENYHlmzy:buhch94) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L174)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Do we really need the nil/empty check on `childrenMetadata` or would the code get an empty array without it? Simplifying this expression if possible would certainly be good.
</blockquote></details>

It's the server's problem now :)
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYtuId4U5seK1Il8jx:-NPrthot-HOZFA96SBWG:b-gitz27) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L175)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't know what I'm talking about, but is `Object.keys` really how you're supposed to iterate over a map?
Is there `childrenMetadata.forEach((name, meta) => ...)`, and then you don't need a lookup inside the closure?
</blockquote></details>

i've found that `.map()` is common in javascript than `forEach()`, though I don't really know why. Regardless, your idea of doing this on the server was great, so it's in Go now!
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYox1N8hsZBbvkqPK6:-NPZ09Di1zRPhabHEM9O:b-5y5cjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L185)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

do such pragmas deserve an explanation?
</blockquote></details>

no, I think they're understood by react-y folks, but I'll defer to David for sure.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYo5FV7iJ_voizVR1Z:-NPZ0LnuBwR6mkOOHI8_:b-rn683w) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L256)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what is this `Loading` business? We looked at `childMetadataRows` above; doesn't that imply everything is already loaded?
</blockquote></details>

No, it doesn't - this function is called on each _render_ cycle, and updated when the dependencies change (i.e. when the API call comes back). So there's an intermediate loading stage we need to handle.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanMetadataTable.tsx` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYt0Ad8XwrVO-lG2oI:-NPrttosBLCZN1V8xp7E:b-rqg696) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanMetadataTable.tsx#L19)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

are `TimestampToMoment`, `ISpanTag` and `RecordingMode` used in this file (I only checked these 3 cause the list seems long to me)? Perhaps I don't know what imports do.
</blockquote></details>

Nope! Copy-paste glitch. Removed, thanks!
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanMetadataTable.tsx` line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYnKro6npDwQaJ2N48:-NPssKwy2SZbNfqMll13:b-6t4k3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanMetadataTable.tsx#L45)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

When I saw the screenshot, I wondered what the circle means in the context of a summary. Please put a tooltip explaining that *some* of the spans are still alive.
</blockquote></details>

done! I also made the circles hollow to indicate something was up. Updated screenshots in the PR description. Thanks!

**Review by dhartunian — COMMENTED** [review]
> Reviewed 1 of 7 files at r1, 3 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97807)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @andreimatei, and @benbardin)

___
*[`-- commits` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPx0EpJErNjw51MTYZ5):*
This commit message needs to conform to our standards. The line is too long and there's no detail about what's being changed.
I think confining this to just a `ui/tracez: ` prefix is nice since it matches the original commits from Andrei.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYox1N8hsZBbvkqPK6:-NPx4HOx076TXW4Qqa4U:b5o5w4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L185)):*
<details><summary><i>Previously, benbardin (Ben Bardin) wrote…</i></summary><blockquote>

no, I think they're understood by react-y folks, but I'll defer to David for sure.
</blockquote></details>

We should justify with a comment when disabling a lint. Generally this linter is correct. Extremely pendantic, but correct :)
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYo5FV7iJ_voizVR1Z:-NPx1EOJC51n7DsKCo_N:b-mf4hc7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L256)):*
<details><summary><i>Previously, benbardin (Ben Bardin) wrote…</i></summary><blockquote>

No, it doesn't - this function is called on each _render_ cycle, and updated when the dependencies change (i.e. when the API call comes back). So there's an intermediate loading stage we need to handle.
</blockquote></details>

I think since this depends on `snapshotLoading` like other sections, they could live within a shared `<Loading>` further up in the hierarchy. This is a bit odd too because the render of this component is already gated on the `childrenMetadata?.length > 0` conditional but also has a `<Loading>` inside. I might be wrong but how will you ever see the error if the conditional isn't true when the error is present?

Ideally we have a top-level `<Loading>` for the data that's in transit to render the page. Once that's done the code assumes that stuff's not null/undefined and proceeds confidently.

**Review by benbardin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97807)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @andreimatei, and @dhartunian)

___
*[`-- commits` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPx0EpJErNjw51MTYZ5):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This commit message needs to conform to our standards. The line is too long and there's no detail about what's being changed.
I think confining this to just a `ui/tracez: ` prefix is nice since it matches the original commits from Andrei.
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYox1N8hsZBbvkqPK6:-NQ2MhJL9s45Fg2euh3e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L185)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

We should justify with a comment when disabling a lint. Generally this linter is correct. Extremely pendantic, but correct :)
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NPYo5FV7iJ_voizVR1Z:-NQ2L67B0p992XM7h85T:b-wqz43t) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d3cd19024bed1579dbbe6a508a84223127bb155/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L256)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I think since this depends on `snapshotLoading` like other sections, they could live within a shared `<Loading>` further up in the hierarchy. This is a bit odd too because the render of this component is already gated on the `childrenMetadata?.length > 0` conditional but also has a `<Loading>` inside. I might be wrong but how will you ever see the error if the conditional isn't true when the error is present?

Ideally we have a top-level `<Loading>` for the data that's in transit to render the page. Once that's done the code assumes that stuff's not null/undefined and proceeds confidently.
</blockquote></details>

Yeah, that's a great point! :) Done.

**Review by dhartunian — APPROVED** [review]
:lgtm: modulo the linter disable comments.

> Reviewed 1 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-:-NQCcqBy8Zb1SoeaNW_N:bkbjm4c)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @andreimatei, and @benbardin)

___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 145 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NQCcljWBzAtTFHmeqk5:-NQCcljWBzAtTFHmeqk6:b-ysczwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/325f0470b44b330e1cc2d05af84f9abba07447e9/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L145)):*
> ```XML
>       spans: spans?.filter(s => s.parent_span_id.equals(spanID)),
>     };
>     // snapshot and spans are a function of nodeID and snapshotID.
> ```

re: these comments, what I'm looking for is a justification of the lint disable. The linter must want to add something that you disagree with, right? otherwise the comment is restating what the array already tells me.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 154 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NQCcmJTAEuAQn-felEW:-NQCcmJTAEuAQn-felEX:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/325f0470b44b330e1cc2d05af84f9abba07447e9/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L154)):*
> ```XML
>       spans: spans?.filter(s => s.span_id.equals(span.parent_span_id)),
>     };
>     // snapshot and spans are a function of nodeID and snapshotID.
> ```

ditto
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 160 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NQCcn3u587c4XfEQvzD:-NQCcn3u587c4XfEQvzE:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/325f0470b44b330e1cc2d05af84f9abba07447e9/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L160)):*
> ```XML
>   const snapshotTime = useMemo(() => {
>     return TimestampToMoment(snapshot?.snapshot.captured_at);
>     // snapshot is a function of nodeID and snapshotID.
> ```

ditto
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 168 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NQCcniF16GYBpOB5B_4:-NQCcniF16GYBpOB5B_5:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/325f0470b44b330e1cc2d05af84f9abba07447e9/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L168)):*
> ```XML
>       return TimestampToMoment(span?.start);
>     },
>     // span are a function of nodeID, snapshotID and spanID.
> ```

ditto

**Review by andreimatei — COMMENTED** [review]
> We're already printing the raw trace! That's in prod now (though the logs are only available for recorded spans)

Right, but in the context of this span page, I'd like to see the logs of this single span, without other spans interleaved.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-:-NQCmY-814GYS88sFCP0:bmyf7gn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @benbardin, and @dhartunian)

**Review by benbardin — COMMENTED** [review]
oh, i see! Let's consider that for later - no need to block this PR on that feature.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-:-NQCnxKa6eWgocRJxE0N:b-zhdent)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @andreimatei, and @dhartunian)

___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 145 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NQCcljWBzAtTFHmeqk5:-NQCo3Zw-3NuvauQFjRU:bh85fkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/325f0470b44b330e1cc2d05af84f9abba07447e9/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L145)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

re: these comments, what I'm looking for is a justification of the lint disable. The linter must want to add something that you disagree with, right? otherwise the comment is restating what the array already tells me.
</blockquote></details>

OK! I'll try to be clearer :) How's this?
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 154 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NQCcmJTAEuAQn-felEW:-NQCo7J82zJlJgPXP1wT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/325f0470b44b330e1cc2d05af84f9abba07447e9/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L154)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 160 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NQCcn3u587c4XfEQvzD:-NQCo8Yh8bRk8efK6mS9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/325f0470b44b330e1cc2d05af84f9abba07447e9/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L160)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 168 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NQCcniF16GYBpOB5B_4:-NQCo8lf1B-3UYs0GT6e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/325f0470b44b330e1cc2d05af84f9abba07447e9/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L168)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97807)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru and @dhartunian)

___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 145 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NQCcljWBzAtTFHmeqk5:-NQSy_ly6ZjuvYKYsPCT:b-atmeu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/325f0470b44b330e1cc2d05af84f9abba07447e9/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L145)):*
<details><summary><i>Previously, benbardin (Ben Bardin) wrote…</i></summary><blockquote>

OK! I'll try to be clearer :) How's this?
</blockquote></details>

I'd like to understand this more, if you don't mind. Can you please spell out for me whose rerendering this `useMemo` saves, and under what circumstances a render would have happened without them? Any chance we're falling in [this situation](https://www.developerway.com/posts/react-re-renders-guide#part5.1), for which `useMemo` is claimed to not help?

**Comment by benbardin** [comment]
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 145 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NQCcljWBzAtTFHmeqk5:-NQVXWZs90F4wTXgVwrx:b-4eslc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/325f0470b44b330e1cc2d05af84f9abba07447e9/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L145)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'd like to understand this more, if you don't mind. Can you please spell out for me whose rerendering this `useMemo` saves, and under what circumstances a render would have happened without them? Any chance we're falling in [this situation](https://www.developerway.com/posts/react-re-renders-guide#part5.1), for which `useMemo` is claimed to not help?
</blockquote></details>

So, the "react system" (waves hands) will call render() on every component whenever the heck it feels like it. Or, more specifically, whenever it thinks it's possible an input value has changed, and in practice more often than that because its understanding is imperfect. This can be when even be when unrelated state changes - React isn't always good at figuring out what is and is not connected, and reasonably errs on the side of re-render.

Most common though is when some props genuinely have changed even while others have not. `snapshotTime` below is a great example of this - we might reasonably expect spanID to change, without nodeID and snapshotID, and we don't necessarily want to re-compute the snapshot's start time. However, with the other calls to useMemo - the trio of `nodeID`, `snapshotID`, `spanID` - you're right that it's unlikely that these three dependencies will remain constant while some other prop changes. In those cases, we're protecting against excessive react framework calls to render().

And what counts as "expensive"? I haven't seen clear, definitive and authoritative documentation on this. In general, consensus seems to be list-based operations count, though. And here that makes intuitive sense - we could have hundreds or thousands of spans, so iterating over all of them could feel long. Is a single TimestampToMoment expensive? Honestly, I don't know; but it's a bunch of weird string manipulation, so my guess is caching this is a net gain.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97807)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[`pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx` line 145 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97807#-NQCcljWBzAtTFHmeqk5:-NQVojPy8jPTA_2NtL4J:b-f3gain) ([raw file](https://github.com/cockroachdb/cockroach/blob/325f0470b44b330e1cc2d05af84f9abba07447e9/pkg/ui/workspaces/cluster-ui/src/tracez/snapshot/spanComponent.tsx#L145)):*
<details><summary><i>Previously, benbardin (Ben Bardin) wrote…</i></summary><blockquote>

So, the "react system" (waves hands) will call render() on every component whenever the heck it feels like it. Or, more specifically, whenever it thinks it's possible an input value has changed, and in practice more often than that because its understanding is imperfect. This can be when even be when unrelated state changes - React isn't always good at figuring out what is and is not connected, and reasonably errs on the side of re-render.

Most common though is when some props genuinely have changed even while others have not. `snapshotTime` below is a great example of this - we might reasonably expect spanID to change, without nodeID and snapshotID, and we don't necessarily want to re-compute the snapshot's start time. However, with the other calls to useMemo - the trio of `nodeID`, `snapshotID`, `spanID` - you're right that it's unlikely that these three dependencies will remain constant while some other prop changes. In those cases, we're protecting against excessive react framework calls to render().

And what counts as "expensive"? I haven't seen clear, definitive and authoritative documentation on this. In general, consensus seems to be list-based operations count, though. And here that makes intuitive sense - we could have hundreds or thousands of spans, so iterating over all of them could feel long. Is a single TimestampToMoment expensive? Honestly, I don't know; but it's a bunch of weird string manipulation, so my guess is caching this is a net gain.


</blockquote></details>

got it, thanks

**Pull request merged**
